### ISSUE TITLE
Questions about mdlstm experiment on IAM dataset

### ISSUE THREAD

**Issue by Marcovaldong** [rwth-i6/returnn#220]
@pvoigtlaender @doetsch I am trying your experiment of mdlstm on IAM, there are some problems about the input and output. 

1. why the image are transformed from 2D tensor to 1D sequence? 
2. what is the meaning of padding 15 pixels from 4 directions? If my own image has size [32, width], is there necessary to do padding? 
3. Why the "classes" is [79, 1]? Im my opinion, the output dimension should be 79 labels + blank = 80.
4. Is there need to padding in a mini-batch, because images in the sam batch have different sizes. Dose the '5 sequence slices' in log 'running 5 sequence slices (1461600 nts) of batch 1059 on device gpu0' means 5 image samples?

**Comment by Marcovaldong** [comment]
@pvoigtlaender  Thank you very much for your reply, I'll check the detail of RETURNN. But I didn't find the place of the network's forward proecedure in the code, Can you tell me where it is? 

Thank you again for your reply.

**Comment by pvoigtlaender** [comment]
With the forward procedure, do you mean the building of the network layers or the code which has a loop around the data to extract the posteriors?

**Comment by Marcovaldong** [comment]
@pvoigtlaender I want to know both things. How does the network is built according to the config, and how does the data flow into the network and flow out of the network, and what size of features after every layer. There is no enough document, and it is difficult to understand these procedure by reading the source code.

**Comment by Marcovaldong** [comment]
@pvoigtlaender Anothe problem is how I can use RASR to decode to get better performance. I noticed that you get your best result by decoding using RASR.

**Comment by pvoigtlaender** [comment]
For building the network, in general, the topology is specified in the json config file and then built accordingly. The specific MDLSTM operation is implemented here: https://github.com/rwth-i6/returnn/blob/master/cuda_implementation/MultiDirectionalTwoDLSTMOp.py
The forwarding should happen (or at least be called from) somewhere in https://github.com/rwth-i6/returnn/blob/master/Engine.py
For the RASR decoding, we did not share the specific config for decoding CTC since it was not "natively" supported and we had to "hack" it a bit to make it work. You'll have to read the DOCs of RASR and it's going to be a bit complicated. I would recommend to consider alternatives like Kaldi which might be easier for CTC decoding.

"In your experiment, can the greedy decoding result of mdlstm be better can CRNN?"
what do you mean with CRNN here? In general, greedy decoding is not going to be too accurate since you use neither a language model nor a lexicon.

**Comment by Marcovaldong** [comment]
@pvoigtlaender   I'm sorry, I didn't express myself clearly. I want to know did you ever compare mdlstm with blstm.

**Comment by pvoigtlaender** [comment]
Not explicitly, but we have prior work with BLSTM here:
P. Voigtlaender, P. Doetsch, S. Wiesler, R. Schluter, and H. Ney, ¨
“Sequence-discriminative training of recurrent neural networks,” in
IEEE International Conference on Acoustics, Speech, and Signal
Processing, Apr. 2015, pp. 2100–2104.

(it's citation [30] in the MDLSTM paper (https://www.vision.rwth-aachen.de/media/papers/MDLSTM_final.pdf) and we also compare to it in the tables)

**Comment by Marcovaldong** [comment]
@pvoigtlaender Thanks for your timely reply, I will read your another paper. There are other two questions I want to ask you, can you help me? I tried to email you, maybe you didn't notice it. 

1. I read the cpp code in [here](https://github.com/rwth-i6/returnn/blob/master/c_support_code_mdlstm.cpp), I want to know how the [float ** datas](https://github.com/rwth-i6/returnn/blob/master/cuda_implementation/c_support_code_mdlstm.cpp#L64) is calculated, I didn't find a place where call this function.

2. I am trying to reproduce your work by PyTorch, I implemented the mdlstm architecture as [here](https://github.com/rwth-i6/returnn/blob/master/cuda_implementation/c_support_code_mdlstm.cpp#L96:L108), which should be theMDLSTM Stable Cell.But this network cannot converge. The structure of my network is the same as yours. Is there some details I didn't noticed. Did you ever try MD LeakyLP Cell?

**Comment by pvoigtlaender** [comment]
1. it's called from lines 512 https://github.com/rwth-i6/returnn/blob/master/cuda_implementation/c_support_code_mdlstm.cpp#L512 and the lines above show how datas is created.
Note that this is a cuda kernel which is launched on the GPU with <<<...>>> notation

**Comment by Marcovaldong** [comment]
I used the weight initialization method by copying your code, The loss decrease from 320+ down to 98, then it cannot be decreased anymore. How can I check the correctness of the gradient?

**Comment by Marcovaldong** [comment]
@pvoigtlaender Thank you very much for your reply, I implemented mdlstm based CRNN, and initialize each layer as your code, now the model can converge successfully.
